### PR TITLE
Threading support

### PR DESCRIPTION
**Pull request by aykevl**
Work in progress. Many necessary parts are not yet implemented, but this gives context for PRs I want to make in the future.

See https://github.com/tinygo-org/tinygo/issues/2446 for motivation (in particular, why this focuses on Linux first) and for a list of things that need to be implemented to make multithreading work in TinyGo.

### CODE REVIEW

**Comment by aykevl** [comment]
Updated after some refactoring (moved internal/llsync into internal/task) and added GC support (stop-the-world scanning by pausing all threads and scanning their stacks).
Main issues remaining are:

 - [x] Some problem with `bits/fcntl.h` that I need to investigate (see CI failures).
 - [x] Timer support.
 - [x] Fixing tests that assume a deterministic order in which goroutines are run.
 - [x] Fixing signal support.

File: src/internal/task/futex_linux.c [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+// import.
+
+#include <stdint.h>
+#include <sys/syscall.h>
+#include <time.h>
+#include <unistd.h>
+
+#define FUTEX_WAIT    0
+#define FUTEX_WAKE    1
+#define FUTEX_PRIVATE 128
```
Comment (leongross) [inline src/internal/task/futex_linux.c]: Why redefine these here if they can be included from `<linux/futex.h>`?

File: src/internal/task/futex_linux.c [hunk 2]
```diff
@@ -0,0 +1,28 @@
+//go:build none
+
+// This file is manually included, to avoid CGo which would cause a circular
+// import.
+
+#include <stdint.h>
+#include <sys/syscall.h>
+#include <time.h>
+#include <unistd.h>
+
```
Comment (leongross) [inline src/internal/task/futex_linux.c]:
```suggestion
#include <unistd.h>
#include <linux/futex.h>
```

File: src/internal/task/futex_linux.c (hunk 1 above)
Comment (leongross) [inline src/internal/task/futex_linux.c]: nit: also afaik the flag itself is called [`FUTEX_PRIVATE_FLAG`](https://elixir.bootlin.com/linux/v6.11/source/include/uapi/linux/futex.h#L26), just to be consistent with [`FUTEX_WAIT`](https://elixir.bootlin.com/linux/v6.11/source/include/uapi/linux/futex.h#L11) and `FUTEX_WAKE`

Comment (aykevl) [reply src/internal/task/futex_linux.c]: Musl (which we use) doesn't provide `linux/futex.h`.

Comment (aykevl) [reply src/internal/task/futex_linux.c]: Thanks! Fixed locally, will be included the next time I push the branch.

**Comment by aykevl** [comment]
Whoa, looks like all tests are passing!

**Comment by jclab-joseph** [comment]
@aykevl Awesome! Thanks for your hard work

**Comment by aykevl** [comment]
I don't think there's a whole lot more I can extract and make separate PRs for, so let's mark this as ready for review!

File: src/internal/task/task_threads.c [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    struct state_pass state = {
+        .start     = (void*)fn,
+        .args      = args,
+        .task      = task,
+        .stackTop  = stackTop,
+    };
+    sem_init(&state.startlock, 0, 0);
+    int result = pthread_create(thread, NULL, &start_wrapper, &state);
+
+    // Wait until the thread has been crated and read all state_pass variables.
```
Comment (eliasnaur) [inline src/internal/task/task_threads.c]: Nit: "created"

File: src/internal/task/task_threads.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+	// and the stop-the-world GC won't see threads that haven't started yet or
+	// are not fully started yet.
+	activeTaskLock.Lock()
+	errCode := tinygo_task_start(fn, args, t, &t.state.thread, &t.state.stackTop)
+	if errCode != 0 {
+		runtimePanic("could not start thread")
+	}
+	t.state.QueueNext = activeTasks
+	activeTasks = t
+	activeTaskLock.Unlock()
```
Comment (eliasnaur) [inline src/internal/task/task_threads.go:116]: Nit: defer the unlock?

File: src/internal/task/task_threads.go [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+	// Wake each paused thread for the second time, so they will resume normal
+	// operation.
+	for t := activeTasks; t != nil; t = t.state.QueueNext {
+		if t != current {
+			t.state.gcSem.Post()
+		}
+	}
+
+	// Allow goroutines to start and exit again.
+	activeTaskLock.Unlock()
```
Comment (eliasnaur) [inline src/internal/task/task_threads.go:212]: Nit: defer?

File: src/internal/task/task_threads.go [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+		runtimePanic("taskExited failed")
+	}
+}
+
+// Futex to wait on until all tasks have finished scanning the stack.
+// This is basically a sync.WaitGroup.
+var scanDoneFutex Futex
+
+// GC scan phase. Because we need to stop the world while scanning, this kinda
+// needs to be done in the tasks package.
```
Comment (eliasnaur) [inline src/internal/task/task_threads.go:149]: FWIW, I find this structure backwards. I would expect the package `runtime` to have `GCScan` that then calls into this package `task`. Especially in light of `runtime.gcScanGlobals` and `runtime.scanStack`.

File: src/runtime/gc_blocks.go [hunk 7]
```diff
@@ -444,7 +452,9 @@ func free(ptr unsafe.Pointer) {
 
 // GC performs a garbage collection cycle.
 func GC() {
+	gcLock.Lock()
 	runGC()
+	gcLock.Unlock()
```
Comment (eliasnaur) [inline src/runtime/gc_blocks.go:458]: Nit: defer?

File: src/runtime/gc_blocks.go [hunk 8]
```diff
@@ -733,6 +743,7 @@ func ReadMemStats(m *MemStats) {
 	m.Sys = uint64(heapEnd - heapStart)
 	m.HeapAlloc = (gcTotalBlocks - gcFreedBlocks) * uint64(bytesPerBlock)
 	m.Alloc = m.HeapAlloc
+	gcLock.Unlock()
```
Comment (eliasnaur) [inline src/runtime/gc_blocks.go:757]: Nit: defer?

**Review by eliasnaur — COMMENTED** [review]
Very much look forward to this work!

File: src/runtime/scheduler.go [hunk 9]
```diff
@@ -27,6 +29,34 @@ func scheduleLogChan(msg string, ch *channel, t *task.Task) {
 	}
 }
 
+func timerQueueAdd(tim *timerNode) {
```
Comment (leongross) [inline src/runtime/scheduler.go]: nit: I think naming the timerNode `tim` is confusing. We should name it either `timer`, `tn`, or just `t`.
I see that name throughout the changes, is it an abbreviation or common term that I am not familiar with?

File: src/internal/task/task_threads.go [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+
+	// Highest address of the stack. It is stored when the goroutine starts, and
+	// is needed to be able to scan the stack.
+	stackTop uintptr
+
+	// Next task in the activeTasks queue.
+	QueueNext *Task
+
+	// Semaphore to pause/resume the thread atomically.
+	pauseSem Semaphore
```
Comment (b0ch3nski) [inline src/internal/task/task_threads.go:30]: Shouldn't `Semaphore` be used with pointer receiver?

File: src/sync/mutex.go [hunk 11]
```diff
@@ -6,131 +6,96 @@ import (
...
+	// Writer lock. Held between Lock() and Unlock().
+	writerLock Mutex
 }
 
-const (
-	rwMutexStateUnlocked = uint32(0)
-	rwMutexStateWLocked  = ^uint32(0)
-	rwMutexMaxReaders    = rwMutexStateWLocked - 1
-)
+const rwMutexMaxReaders = 1 << 30
```
Comment (leongross) [inline src/sync/mutex.go:32]: Is there a reason you chose this specific value?

File: src/internal/task/task_threads.c (hunk 3 above)
Comment (aykevl) [reply src/internal/task/task_threads.c]: Thanks, fixed!

File: src/internal/task/task_threads.go (hunk 4 above)
Comment (aykevl) [reply src/internal/task/task_threads.go:116]: I avoid `defer` in places like these since:

1. `defer` can cause heap allocations in some cases (mainly in loops, so not really relevant here, but it's not very visible).
2. Using `defer` currently makes every call in the function slower and larger, since it needs to be able to recover.
3. We don't need the run-defer-on-panic behavior here: if we panic, something is very wrong and it doesn't help to unlock.

Admittedly it _could_ be used here, but in many other places (e.g. `GCScan`) it can lead to unexpected side effects if it allocates.

So in short, `defer` as it is currently implemented is inefficient and can have unexpected side effects so I avoid its use in the runtime.

(Same goes for all the other comments regarding `defer`).

File: src/internal/task/task_threads.go (hunk 6 above)
Comment (aykevl) [reply src/internal/task/task_threads.go:149]: I agree with you, and I first tried to put the equivalent of `GCScan` in the runtime. But `GCScan` also needs access to various private symbols like `activeTaskLock`, `activeTasks`, `scanDoneFutex`, etc so to avoid exporting many private details of the tasks package I put it there.

I can however make `scanCurrentStack` and `gcScanGlobals` callbacks, so they don't directly call back to the runtime. But then I don't know what to do with `scanCurrentStack` in `tinygo_task_gc_pause`. In the end, the GC and the scheduler do need some tight integration, I can't see a way around that to be honest.

File: src/internal/task/task_threads.go (hunk 10 above)
Comment (aykevl) [reply src/internal/task/task_threads.go:30]: No. The `state` struct (or actually the `Task` struct where it is embedded) will always be accessed via a pointer, like `*Task`. So essentially when calling a function like `pauseSem.Post()`, what happens is something like `SemaphorePost(&t.state.pauseSem)`.

Also, I think you mean to say that `pauseSem` should have the type `*Semaphore`? Since the methods on the `Semaphore` type do in fact have a pointer receiver:

```go
func (s *Semaphore) Post() { ... }
func (s *Semaphore) Wait() { ... }
```

File: src/sync/mutex.go (hunk 11 above)
Comment (aykevl) [reply src/sync/mutex.go:32]: No reason, except that it's the same value as used in the [upstream `RWMutex` implementation](https://github.com/golang/go/blob/master/src/sync/rwmutex.go#L47):

```go
const rwmutexMaxReaders = 1 << 30
```

File: src/runtime/scheduler.go (hunk 9 above)
Comment (aykevl) [reply src/runtime/scheduler.go]: Good point, I've updated the naming of some variables here to be more consistent.
(I think `tim` is just short for `timer`).

**Comment by aykevl** [comment]
Updated the PR after some review comments. And rebased on dev to incorporate #4435.

File: src/internal/task/task_threads.go (hunk 10 above)
Comment (b0ch3nski) [reply src/internal/task/task_threads.go:30]: My point was about a pointer (pun intended :smile:) on `*Semaphore`. While `Semaphore` methods do have pointer receiver, the instances of `pauseSem` and `gcSem` members would get copied when `state` struct instance is ever passed around.

I've created playground snippet which uses stdlib `sync.Mutex` as an example to even trigger `go vet` warnings: https://go.dev/play/p/cV7h16_JS1R

Comment (aykevl) [reply src/internal/task/task_threads.go:30]: > While `Semaphore` methods do have pointer receiver, the instances of `pauseSem` and `gcSem` members would get copied when `state` struct instance is ever passed around.

That's why it should never be copied around!
Because it's an internal API, we can make sure this is the case. (And it is).

File: src/sync/mutex.go (hunk 11 above)
Comment (leongross) [reply src/sync/mutex.go:32]: Sounds reasonable enough to me.

**Comment by aykevl** [comment]
Rebased (which was a bit of work after #4385 got merged), let's see whether CI is happy with this.

**Comment by aykevl** [comment]
Added some comments to clarify stuff! Ready for review!

(Waiting for this, since lots of people want multicore support on the RP2040 and this PR is a prerequisite).

File: src/runtime/gc_blocks.go [hunk 12]
```diff
@@ -492,6 +501,10 @@ func runGC() (freeBytes uintptr) {
 		finishMark()
 	}
 
+	// If we're using threads, resume all other threads before starting the
+	// sweep.
+	gcResumeWold()
```
Comment (eliasnaur) [inline src/runtime/gc_blocks.go]: nit: typo: `gcResumeWorld`

**Review by eliasnaur — COMMENTED** [review]
LGTM, but I'm certainly no expert in this area.

File: src/runtime/gc_blocks.go (hunk 12 above)
Comment (aykevl) [reply src/runtime/gc_blocks.go]: Oops, fixed!
That's the danger of autocomplete (it makes it very easy to repeat a single typo).

**Comment by aykevl** [comment]
I'd love to get this merged, since it blocks a few other things I'm working on (including the much-demanded RP2040 multicore support).

**Comment by b0ch3nski** [comment]
LGTM. I've read the code and there is nothing in there I could be worried about, although I understand mostly the general idea of this.

**Comment by deadprogram** [comment]
Thanks for all of the useful comments in this PR @aykevl it has made it a lot easier to review. Amazing work!

Thank you very much to @leongross @jclab-joseph @eliasnaur @b0ch3nski for all of the thoughtful comments and feedback!

Now merging :rocket:

**Comment by jfcg** [comment]
Amazing !
Are goroutines mapped to OS-threads one-to-one ?
Or is there a multiplexer like in gc toolchain?
Great work, thanks..

**Comment by aykevl** [comment]
Goroutines are mapped one-to-one to OS threads for ease of implementation. It really simplifies the implementation a lot.
